asic/src/tofino_asic/ports.rs

```diff
@@ -412,6 +428,118 @@ pub fn set_prbs(
     Ok(())
 }
 
+const MAX_N_LANES: usize = 16;
+
+const fn empty_ber() -> bf_port_ber_t {
+    bf_port_ber_t {
+        ber: bf_port_ber_t_ber_union_ {
+            tof2: bf_tof2_ber_t {
+                ber_aggr: 0.0,
+                ber_per_lane: [0.0; MAX_N_LANES],
+                sym_err_ctrs: [0; MAX_N_LANES],
+            },
+        },
+    }
+}
+
+// Helper struct to return the lanes in the `bf_tof2_ber_t` array fields that
+// are expected to be non-zero. This is based entirely on the speed of the link.
+// See
+// https://github.com/oxidecomputer/tofino-sde/blob/0e4cfe10c92db223a953923d77f09a299b82157b/pkgsrc/bf-drivers/src/bf_pm/pm.c#L2120-L2140
+// for the logic.
+struct BerLanes {
+    start: usize,
+    end: usize,
+}
+
+impl BerLanes {
+    const fn all() -> Self {
+        Self {
+            start: 0,
+            end: MAX_N_LANES - 1,
+        }
+    }
+
+    // Return the lanes used for BER counters for the provided speed.
+    //
+    // Use all the counters if we can't reliably determine which ones to read.
+    fn new(hdl: PortHdl, speed: PortSpeed) -> Self {
+        // This magic logic appears in `pm_port_tof2_ber_get()`, where it
+        // appears to be used to determine the right counters to read from the
+        // serdes firmware.
+        const MASK: u64 = 0x07;
+        const FACTOR: u64 = 2;
+        let Ok(start) = usize::try_from((u64::from(hdl) & MASK) * FACTOR)
+        else {
+            return Self::all();
+        };
+        match speed {
```

Comment:
This calculation doesn't seem complete.  Can't we build a 100g port from 10 10g links?  40g from 4 10g links?  There is a `serdes::lane_count()` function that will return the actual number of lanes configured.

Reply:
I agree that you _could_, but I'm not sure we support that today. Don't we or the SDE automatically figure out the number of lanes? That's what `PhysPort::allocate_channels_for()` is doing, anyway.

I'm also trying to replicate exactly what the SDE does internally, since I have very little insight into what the various counters actually mean. The other option is to ignore all this, and return 16 of them all the time. I dislike that a bit because then it's on the operator to figure out if zero means "this is fine" or "this wasn't used".

Reply:
I think I see a better way to do this, in any case, and in a way that avoids needing the speed at all. We know the number of _lanes_, as you point out, and we can replicate the logic in the SDE with just ASIC port identifier. I.e., we don't need the speed at all. I'll rework all this.

Reply:
Done in [5ce22d7](https://github.com/oxidecomputer/dendrite/pull/20/commits/5ce22d7c672cf3fcffce5bda96cb4f2871d766f6)